### ISSUE TITLE
Dramatically worse performance for LIMIT comparing to vanilla postgres

### ISSUE THREAD

**Issue by Nk185** [timescale/timescaledb#1879]
**Relevant system information:**
 - OS: Windows 10.0.17134.885 (1803/April2018Update/Redstone4)
 - PostgreSQL version (output of `postgres --version`): 11.7
 - TimescaleDB version (output of `\dx` in `psql`): 1.7.0

**Describe the bug**
As said in the title, TimescaleDB perform much worse than vanilla Postgres for queries like `SELECT x FROM y [WHERE z = 'Z'] ORDER BY j DESC LIMIT 1` (the '[]' means optional part of query - presence or absence doesn't affect performance much) ran against hypertables. In fact, the higher number of chunks, the worse performance yielded by TimescaleDB.

**To Reproduce**
1. Create a table "vanilla" with columns id, time, payload.
2. Create a btree index on the "time" column with desc sorting.
3. Create a table "ts" with columns id, time, payload.
4. Convert the "ts" table to hypertable with chunk interval 1 day (partitioned by the "time" column, obviously)
5. Into both tables insert the same amount of data. In my case, it was 124 350 021 records spread over 1001 days (approximately 124 225 records per day)
6. Run query `SELECT payload FROM vanilla ORDER BY time DESC LIMIT 1`. Note execution time.
7. Run query `SELECT payload FROM ts ORDER BY time DESC LIMIT 1`. Note execution time.
8. Compare execution times.

**Expected behavior**
Performance the same as for (or at least close to) the vanilla Postgres.

**Actual behavior**
Drastically worse performance for TimescaleDB comparing to the vanilla Postgres.

**Additional context**
My query and execution plan [for TimescaleDB](https://gist.github.com/Nk185/316c7ed68290864d4b9fd95aabfe8da7), and [for vanilla Postgres](https://gist.github.com/Nk185/2e1bf66737a9f44026ee15762e502ec0).

It looks like this happens like so because TimescaleDB involves all the chunks of a hypertable. This also the reason why `[WHERE z = 'Z']` (from the description paragraph) makes no difference - if there are 1000 chunks for hypertable `y` and 'Z' present only in, say, 2 of them, TimescaleDB anyway gonna involve all 1000 even though there might be an idx on `z` and `j`. So the more chunks you have the worst performance you will get.

The issue possibly relates to #1489. It could be that the root cause is the same for continuous aggregates too as the symptoms look similar.

**Comment by svenklemm** [comment]
@Nk185 can you provide EXPLAIN output for the offending queries.

**Comment by Nk185** [comment]
@svenklemm have you checked links from the "Additional context" paragraph? It refers to execution plans from EXPLAIN as well as SQL queries for both timescale and vanilla Postgres.

**Comment by Nk185** [comment]
@svenklemm given my comment above, do you have enough information from my side to proceed? Or you are still waiting for something from my side?

**Comment by svenklemm** [comment]
@Nk185 looks like your chunk_time_interval should be chosen much bigger. Your query does not have any constraints on time so alls chunks have to be considered during planning leading to this very high planning time as you got so many chunks.

**Comment by Nk185** [comment]
@svenklemm 

>looks like your chunk_time_interval should be chosen much bigger.

Well, it might help but won't solve the issue - the number of chunks will be monotonically growing over time, so eventually, I will end up with the same problem sooner or later.  Plus, the current value for the interval follows best practices outlined you Timescaledb guide.

So changing the value is not an option to go.

>Your query does not have any constraints on time, so alls chunks have to be considered during planning

It's not intended to have one - the point of this query is to bring back last N observations regardless of when they were observed. Furthermore, as a caller, I don't know what time interval I'm interested in; neither it's the consumer's responsibility to take care of handling this (IMO).

On the other hand, on the DB side, you can use chunk/index/table metadata to estimate which chunks have relevant data and how much data they have to satisfy the LIMIT clause.

To handle these "last observation queries," we introduced a "metadata" table with most of the information used in the WHERE clause and timestamp of the last record. So that to take the last observation, we have abracadabra in the form of stored procedure like the one below; thus, inner SELECT always gets a fine execution plan with a single chunk. But this solution has a bad smell (and a lot of overhead), and it won't work when you need more than one item (e.g., you need the last 10 observations). Nothing will work actually
```
get_last_xxxxxxxxx_by_id(IN _ids uuid[])
    RETURNS table
            (
                id       uuid,
                x        varchar(200),
                y        double precision,
                z        integer,
                time_utc timestamp WITHOUT TIME ZONE
            )
AS
$$
DECLARE
    _matching_metadata_entry record;
BEGIN
    FOR _matching_metadata_entry IN SELECT m.id, m.x, m.time_utc
                               FROM xxxxxxxxx_metadata m
                               WHERE m.id = ANY (_ids)
        LOOP
            SELECT _matching_metadata_entry.id,
                   _matching_metadata_entry.x,
                   xs.ts_data_y,
                   xs.ts_data_z,
                   _matching_metadata_entry.time_utc
            INTO id,
                x,
                y,
                z,
                time_utc
            FROM xxxxxxxxx xs
            WHERE xs.id = _matching_metadata_entry.id
              AND xs.time_utc = _matching_metadata_entry.time_utc;

            RETURN NEXT;

        END LOOP;
END;
$$
    LANGUAGE plpgsql
    STABLE;
```

**Comment by adam-bhaiji** [comment]
Just to add further to @Nk185 comment, even if you do have a time filter on the created buckets, the query is extraordinarily slow and often hangs. I would hazard a guess that all chunks are being considered, as outlined at the start of this issue in https://github.com/timescale/timescaledb/issues/1489

**Comment by mralexes** [comment]
Possibly related to #2897

**Comment by NunoFilipeSantos** [comment]
@Nk185 can you please try with our latest TimescaleDB 2.4.2 (PG 12 or PG13) and get back to us?

**Comment by NunoFilipeSantos** [comment]
Hi @Nk185! Closing for now. When you can get back to us, please feel free to reopen the issue.

**Comment by behappyall** [comment]
Hi @NunoFilipeSantos! This issue is still relevant. I was able to reproduce it with the latest TimescaleDB 2.5.1 (PG 14).

**Comment by NunoFilipeSantos** [comment]
Thank you @behappyall, reopening.

**Comment by akuzm** [comment]
The problem is in slow planning:
```
    "Planning Time": 828.856,
    "Triggers": [],
    "Execution Time": 8.713
```

**Comment by akuzm** [comment]
The slow planning was fixed in 2.6: https://github.com/timescale/timescaledb/issues/2897
